File: internal/core/kernel/controller.go
package kernel

import (
	"context"

	"mvmctl/internal/lib/model"
)

// Controller matches Python's Controller.
type Controller struct {
	kernel *model.KernelItem
	repo   Repository
}

// NewController creates a Controller bound to a resolved kernel.
func NewController(kernel *model.KernelItem, repo Repository) *Controller {
	return &Controller{kernel: kernel, repo: repo}
}

// Get returns the bound kernel item.
func (c *Controller) Get() *model.KernelItem {
	return c.kernel
}

// SetDefault sets this kernel as the default.
func (c *Controller) SetDefault(ctx context.Context) error {
	return c.repo.SetDefault(ctx, c.kernel.ID)
}

File: legacy/python/src/mvmctl/core/kernel/_controller.py
"""Kernel controller — stateful kernel operations bound to a single instance."""

from __future__ import annotations

from mvmctl.core.kernel._repository import KernelRepository
from mvmctl.core.kernel._resolver import KernelResolver
from mvmctl.models import KernelItem


class KernelController:
    """
    Stateful kernel controller — bound to a single kernel instance.

    Args:
        entity: KernelItem or string identifier (ID prefix) to resolve.
        repo: KernelRepository for DB operations.

    """

    def __init__(
        self, entity: str | KernelItem, repo: KernelRepository
    ) -> None:
        self._repo = repo
        if isinstance(entity, KernelItem):
            self._kernel = entity
        else:
            resolver = KernelResolver(repo)
            self._kernel = resolver.resolve(entity)

    def get(self) -> KernelItem:
        """Return the bound kernel item."""
        return self._kernel

    def set_default(self) -> None:
        """Set this kernel as the default."""
        self._repo.set_default(self._kernel.id)
